python/lsst/ts/tcpip/client.py

```diff
@@ -0,0 +1,262 @@
+        | None = None,
+        monitor_connection_interval: float = DEFAULT_MONITOR_CONNECTION_INTERVAL,
+        name: str = "",
+        **kwargs: typing.Any,
+    ) -> None:
+        self.host = host
+        self.port = port
+        self.log = log.getChild(f"{type(self).__name__}({name})")
+        self.__connect_callback = connect_callback
+        self.monitor_connection_interval = monitor_connection_interval
+        self.name = name
+
+        # Has the connection been made and not closed at this end?
+        self.should_be_connected = False
+
+        # Has start been called? Used to prevent calling it more than once.
+        self._started = False
+
+        # Was the client connected last time `call_connected_callback`
+        # called? Used to prevent multiple calls to ``connect_callback``
+        # for the same connected state.
+        self._was_connected = False
+
+        self._monitor_connection_task: asyncio.Future = asyncio.Future()
+        self._monitor_connection_task.set_result(None)
+
+        self.reader: asyncio.StreamReader | None = None
+        self.writer: asyncio.StreamWriter | None = None
+
+        # Task which is set done when client connects.
+        self.start_task: asyncio.Future = asyncio.create_task(self.start(**kwargs))
+
+        # Task which is set done when the client is closed.
+        self.done_task: asyncio.Future = asyncio.Future()
+
+    @property
+    def connected(self) -> bool:
+        """Return True if this client is connected to a server."""
+        return not (
+            self.reader is None
+            or self.writer is None
+            or self.reader.at_eof()
+            or self.writer.is_closing()
+        )
+
+    async def call_connect_callback(self) -> None:
+        """Call self.__connect_callback.
+
+        This is always safe to call. It only calls the callback function
+        if that function is not None and if the connection state has changed
+        since the last time this method was called.
+        """
+        connected = self.connected
+        self.log.debug(
+            f"call_connect_callback: {connected=}; was_connected={self._was_connected}"
+        )
+        if self._was_connected != connected:
+            self._was_connected = connected
+            if self.__connect_callback is not None:
+                try:
+                    await self.__connect_callback(self)
+                except Exception:
+                    self.log.exception("connect_callback failed.")
+
+    async def basic_close(self) -> None:
+        """Close the connected client socket, if any.
+
+        Like `close` except does not clear ``self.should_be_connected``,
+        nor cancel ``self._monitor_connection_task``.
+        """
+        try:
+            self.log.info("Closing")
+            if self.writer is None:
+                return
+
+            writer = self.writer
+            self.writer = None
+            await utils.close_stream_writer(writer)
```

Comment:
Why not just: `await utils.close_stream_writer(self.writer)`?

I guess I do not really understand the logic here.

Reply:
This sets self.writer is set to None immediately, while the writer is being closed. It means that `connected` will return False as soon as possible.

It also assures that the writer will be set to None regardless of whether closing succeeds or fails (one could use try/finally for that, but this method has the other advantage I just mentioned, and is more efficient).

Another common way to spell this pattern is:
```
self.writer, writer = None, self.writer
...do something with writer
```